composer/trainer/checkpoint.py

```diff
@@ -139,9 +172,24 @@ def save_checkpoint(self, state: State, seed: int, device: Device, config: Optio
             'rng': self._get_rng_state(device=device),  # stored across all ranks
             'seed': ddp.all_gather_object(seed),
         }
+
+        if self.save_event == Event.EPOCH_END:
+            tag = f"ep{state.epoch}"
+        elif self.save_event == Event.BATCH_END:
+            tag = f"it{state.step}"
+        else:
+            raise ValueError(f"Invalid checkpoint event: {self.save_event}")
+
+        if state.model.__class__.__qualname__ == 'DeepSpeedEngine':
+            state.model.save_checkpoint(self.checkpoint_folder, tag)  # type: ignore
+
         if ddp.get_global_rank() != 0:
             # only rank 0 saves checkpoints
             # Need the check down here so all the DDP syncs will work for generating the checkpoint
+
+            # Sync before exiting so that even the non rank 0 processes cannot exit before the
```

Comment:
Good catch, thanks! Generally, I'd prefer this style of short circuiting early to reduce indentation. But, since all processes need to have the same # of ddp barriers, I personally find it easier to follow if there is just one line to `ddp.barrier` that is shared by both the rank-0 and non rank-0 processes...e.g. something like:

```python3
if ddp.get_global_rank() != 0:
        checkpoint_filepath = get_mosaic_checkpoint_filepath(self.checkpoint_folder, tag)
        if not os.path.exists(os.path.join(self.checkpoint_folder, tag)):
            os.makedirs(os.path.join(self.checkpoint_folder, tag))
        with open(checkpoint_filepath, 'xb') as f:
              ....
ddp.barrier()  # <- same line is used for all code paths
```